File: ysw/v1/Data_preparation/ex.py
# ref
# https://en.wikipedia.org/wiki/Rotation_matrix#Quaternion
# https://en.wikipedia.org/wiki/Quaternions_and_spatial_rotation#Quaternion-derived_rotation_matrix
# https://github.com/MichaelGrupp/evo/issues/153
# https://github.com/MichaelGrupp/evo/blob/master/evo/core/transformations.py#L1258

import numpy
import math

_EPS = numpy.finfo(float).eps * 4.0
print(_EPS) # 8.881784197001252e-16

def quaternion_matrix(quaternion):
    """Return homogeneous rotation matrix from quaternion.

    >>> M = quaternion_matrix([0.99810947, 0.06146124, 0, 0])
    >>> numpy.allclose(M, rotation_matrix(0.123, [1, 0, 0]))
    True
    >>> M = quaternion_matrix([1, 0, 0, 0])
    >>> numpy.allclose(M, numpy.identity(4))
    True
    >>> M = quaternion_matrix([0, 1, 0, 0])
    >>> numpy.allclose(M, numpy.diag([1, -1, -1, 1]))
    True

    """
    q = numpy.array(quaternion, dtype=numpy.float64, copy=True)
    n = numpy.dot(q, q)
    if n < _EPS:
        return numpy.identity(4)
    q *= math.sqrt(2.0 / n)
    q = numpy.outer(q, q)
    return numpy.array([
        [1.0-q[2, 2]-q[3, 3],     q[1, 2]-q[3, 0],     q[1, 3]+q[2, 0]],
        [    q[1, 2]+q[3, 0], 1.0-q[1, 1]-q[3, 3],     q[2, 3]-q[1, 0]],
        [    q[1, 3]-q[2, 0],     q[2, 3]+q[1, 0], 1.0-q[1, 1]-q[2, 2]]])

test = [1, 0, 0, 0]
test_q = numpy.array(test, dtype=numpy.float64, copy=True)
test_n = numpy.dot(test_q, test_q)  # test_n: 1
test_q *= math.sqrt(2.0 / test_n)
print(test_q)   # [1.41421356 0.         0.         0.        ] ????
test_q = numpy.outer(test_q, test_q)
print(test_q)
# [[2. 0. 0. 0.]
#  [0. 0. 0. 0.]
#  [0. 0. 0. 0.]
#  [0. 0. 0. 0.]]

# 100th line of traj.txt created by faster-LIO with first bag file; q_x, q_y, q_z, q_w
# change order to [q_w, q_x, q_y, q_z]
quat = [0.999390925644008, -0.008357784211309, -0.004638522929725, -0.033562021520099]
rot = quaternion_matrix(quat)
print('\n'*3, rot)
    # [[ 0.99770415  0.0671607  -0.00871039]
    #  [-0.06700562  0.99760748  0.01701674]
    #  [ 0.0098324  -0.01639403  0.99981726]]

# check : https://www.andre-gaschler.com/rotationconverter/

File: ysw/v1/Data_preparation/ex.cc
// cpp version of quaternion_matrix() in https://github.com/MichaelGrupp/evo/blob/master/evo/core/transformations.py#L1258

#include <iostream>
#include <vector>
#include <cmath>
#include <iomanip>

const double _EPS = 1e-15; // Small threshold for near-zero quaternion magnitudes

std::vector<std::vector<double>> quaternion_matrix(const std::vector<double>& quat) {
    std::vector<double> q = quat;
    
    // Calculate quaternion magnitude squared
    double n = q[0]*q[0] + q[1]*q[1] + q[2]*q[2] + q[3]*q[3];
    
    // Return identity for near-zero quaternions
    if (n < _EPS) {
        return {
            {1,0,0},
            {0,1,0},
            {0,0,1}
        };
    }
    
    // Normalization factor
    double scale = sqrt(2.0 / n);
    for (auto& val : q) val *= scale;

    // Calculate outer product matrix (not vector product)
    std::vector<std::vector<double>> outer(4, std::vector<double>(4));
    for (int i = 0; i < 4; ++i)
        for (int j = 0; j < 4; ++j)
            outer[i][j] = q[i] * q[j];
            
    return {
        {1.0 - outer[2][2] - outer[3][3], outer[1][2] - outer[3][0], outer[1][3] + outer[2][0]},
        {outer[1][2] + outer[3][0], 1.0 - outer[1][1] - outer[3][3], outer[2][3] - outer[1][0]},
        {outer[1][3] - outer[2][0], outer[2][3] + outer[1][0], 1.0 - outer[1][1] - outer[2][2]}
    };
}

int main() {
    std::vector<double> quat = { // w, x, y, z
        0.999390925644008,
        -0.008357784211309,
        -0.004638522929725,
        -0.033562021520099
    };

    auto rot = quaternion_matrix(quat);

    std::cout << std::fixed << std::setprecision(8);
    for (const auto& row : rot) {
        for (double val : row) {
            std::cout << std::setw(12) << val << " ";
        }
        std::cout << "\n";
    }

    return 0;
}
